mypy/checkmember.py

```diff
@@ -496,3 +520,82 @@ def map_type_from_supertype(typ: Type, sub_info: TypeInfo,
     # in inst_type they are interpreted in subtype context. This works even if
     # the names of type variables in supertype and subtype overlap.
     return expand_type_by_instance(typ, inst_type)
+
+
+F = TypeVar('F', bound=FunctionLike)
+
+
+def bind_self(method: F, actual_self: Type = None) -> F:
+    '''Return a copy of `method`, with the type of its first parameter (usually
+    self or cls) bound to actual_self.
+
+    If the type of `self` is a generic type (T, or Type[T] for classmethods),
+    instantiate every occurrence of type with actual_self in the rest of the
+    signature and in the return type.
+
+    actual_self is the type of E in the expression E.copy(). It is None in
+    compatibility checks. In this case we treat it as the erasure of the
+    declared type of self.
+
+    This way we can express "the type of self". For example:
+
+    T = TypeVar('T', bound='A')
+    class A:
+        def copy(self: T) -> T: ...
+
+    class B(A): pass
+
+    b = B().copy()  # type: B
+
+    '''
+    if isinstance(method, Overloaded):
+        return cast(F, Overloaded([bind_self(c, method) for c in method.items()]))
+    assert isinstance(method, CallableType)
+    func = method
+    if not func.arg_types:
+        # invalid method. return something
+        return cast(F, func)
+    if func.arg_kinds[0] == ARG_STAR:
+        # The signature is of the form 'def foo(*args, ...)'.
+        # In this case we shouldn'func drop the first arg,
+        # since func will be absorbed by the *args.
+
+        # TODO: infer bounds on the type of *args?
+        return cast(F, func)
+    self_param_type = func.arg_types[0]
+    if func.variables and (isinstance(self_param_type, TypeVarType) or
+            (isinstance(self_param_type, TypeType) and
+             isinstance(self_param_type.item, TypeVarType))):
+        if actual_self is None:
+            # Type check method override
+            # XXX value restriction as union?
+            actual_self = erase_to_bound(self_param_type)
+
+        typearg = infer_type_arguments([x.id for x in func.variables],
+                                       self_param_type, actual_self)[0]
+
+        def expand(target: Type) -> Type:
+            return expand_type(target, {func.variables[0].id: typearg}, False)
```

Comment:
Currently none that I know of, but I'm prototyping a tool that will need it for runtime type checking. They way it's used is pretty complex so you have to take my word for it as the margin isn't wide enough for a good explanation :-)